governator-core/src/main/java/com/netflix/governator/ScanningModuleBuilder.java

```diff
@@ -0,0 +1,231 @@
+                                            }
+                                        }
+                                    } catch (Exception e ) {
+                                        addError("Unable to scan '%s'. %s", file.getCanonicalPath(), e.getMessage());
+                                        addError(e);
+                                    }
+                                } else {
+                                    DirectoryClassFilter filter = new DirectoryClassFilter(classLoader);
+                                    for ( String className : filter.filesInPackage(url, basePackage) ) {
+                                        try (InputStream is = filter.bytecodeOf(className)) {
+                                            handleClass(is);
+                                        }
+                                    }
+                                }
+                            } catch (Exception e) {
+                                addError("Unable to scan jar '%s'. %s ", url, e.getMessage());
+                                addError(e);
+                            }
+                        }
+                    } catch ( Exception e ) {
+                        addError("Classpath scanning failed for package \'" + basePackage + "\'");
+                        addError(e);
+                    }
+                }
+            }
+            
+            private boolean isJarURL(URL url) {
+                String protocol = url.getProtocol();
+                return "zip".equals(protocol) || "jar".equals(protocol) ||
+                        ("file".equals(protocol) && url.getPath().endsWith(".jar"));
+            }
+            
+            private void handleClass(InputStream inputStream) throws IOException {
+                new ClassReader(inputStream).accept(new ClassVisitor(Opcodes.ASM5) {
+                    private String className;
+                    
+                    @Override
+                    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+                        className = name.replace('/', '.');
+                        super.visit(version, access, name, signature, superName, interfaces);
+                    }
+                    
+                    @Override
+                    public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
+                        Type type = getType(desc);
+                        if (includeRule.apply(className)) {
+                            for (AnnotatedClassScanner scanner : scanners) {
+                                if (getType(scanner.annotationClass()).equals(type)) {
+                                    try {
+                                        Class<?> cls = Class.forName(className, false, classLoader);
+                                        scanner.applyTo(binder(), cls.getAnnotation(scanner.annotationClass()), Key.get(cls));
```

Comment:
maybe defer class loading to the scanner; pass it the classname and the binder only